### PR TITLE
retraining with MELD

### PR DESCRIPTION
**Pull request by amith-ananthram**
added 4 files to allow us to re-train the vox speaker id reference model to do emotion detection with the meld corpus.  

1) prepare_meld.sh - creates the expected directory structure for meld input / output data, modifies the reference model for retraining (changing the output dimensionality of the final layer, fixing the learning rates of the first 6 layers), and generates inputs for the retraining script (in the same format that v2/run.sh does)

2) generate_meld_inputs.py - takes the manifest files and directory structure of the meld corpus and generates utt2spk, spk2utt and wav.scp files for each of dev / train / test; note that, instead of using the speaker identifications from the MELD corpus we instead use the emotion labels (to allow us to re-use as much of v2/run.sh as possible); also, as MELD has a different sampling rate than vox2, we converting to .wav we change the sampling rate so that they match

3) rerun_with_meld.sh - borrows heavily from run.sh; prepares the MELD corpus, augments with noise/speech/music/reverb, generates MFCCs, removes silence, filters out utterances that are too short, trains the model (note that it saves all intermediate and final data in a separate file structure from run.sh so the two can be run w/in the same namespace wo stepping on each others' toes)

4) run_meld_xvector.sh - borrows heavily from run_xvector.sh, generates neural network training examples and trains the network

### CODE REVIEW

**Comment by amith-ananthram** [comment]
(the next PR will include a README that outlines the steps for retraining + a copy of vox2/run.sh modified for the purpose of retraining)

File: egs/voxceleb/v2/generate_meld_inputs.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+def generate_utt2spk(utterances, output_data_dir):
+	with open(os.path.join(output_data_dir, UTT2SPK_FILE), 'w') as f:
+		for utterance in sorted(utterances, key=lambda utterance: utterance.get_id()):
+			f.write("%s %s\n" % (utterance.get_id(), EMOTION_TO_ID[utterance.emotion]))
+
+def generate_wavscp(utterances, input_data_dir, output_data_dir):
+	with open(os.path.join(output_data_dir, WAV_FILE), 'w') as f:
+		for utterance in sorted(utterances, key=lambda utterance: utterance.get_id()):
+			mp4_file_path = os.path.join(input_data_dir, utterance.get_filename())
+			wav_creation_cmd = "ffmpeg -v 8 -i %s -f wav -ar 16000 -acodec pcm_s16le -|" % (mp4_file_path)
```
Comment (amith-ananthram) [inline egs/voxceleb/v2/generate_meld_inputs.py:79]: as MELD has a different sample rate than vox, i specify the vox sample rate when converting to wav here

File: egs/voxceleb/v2/prepare_meld.sh [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+	#
+	# END: copied with minimal modification from run_xvector.sh
+	#
+
+	# now, with our updated target config in hand, we copy our
+	# reference model, modifying its final output layer and setting
+	# the learning rates for the first 6 layers to 0
+	nnet3-copy \
+		--nnet-config="$MODEL_OUTPUT_DIR/modified_nnet.config" \
```
Comment (amith-ananthram) [inline egs/voxceleb/v2/prepare_meld.sh:106]: going through the Kaldi code, it seems like specifying a -nnet-config when copying a nnet will result in kaldi throwing away any layers that are specified in both the base model and the config (preferring whatever is specified in the config) but this doesn't seem to be happening in practice.  when i get around to retraining the modified reference model on emotion, it fails with:
```
ERROR (nnet3-train[5.5.528~1-880df]:ComputeObjectiveFunction():nnet-training.cc:349) Nnet versus example output dimension (num-classes) mismatch for 'output': 3 (nnet) vs. 6 (egs)
```
(where 3 is the number of Vox speakers that I trained the reference model on).  Should I be doing this some other way?

File: egs/voxceleb/v2/rerun_with_meld.sh [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    ${data_dir}/train_combined ${data_dir}/train_combined_no_sil ${root}/exp/train_combined_no_sil
+  utils/fix_data_dir.sh ${data_dir}/train_combined_no_sil
+fi
+
+# ./run.sh does a bunch of filtering of utterances by speakers
+# that are too infrequent -- we skip that here speaker=emotion label 
+if [ $stage -eq 7 ]; then
+  # Now, we need to remove features that are too short after removing silence
+  # frames.  We want atleast ~1s (100 frames) per utterance. (note this is smaller than in v2/run.sh)
+  min_len=100
```
Comment (amith-ananthram) [inline egs/voxceleb/v2/rerun_with_meld.sh:144]: note that this is a change from the origin v2/run.sh script which filters anything shorter than 5s.  as our model is already partially trained off Vox and we don't have a ton of labeled emotion data, I allow shorter MELD segments to remain part of our training set.

File: egs/voxceleb/v1/local/nnet3/xvector/run_meld_xvector.sh [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    utils/create_split_dir.pl \
+     /export/b{03,04,05,06}/$USER/kaldi-data/egs/voxceleb2/v2/xvector-$(date +'%m_%d_%H_%M')/$egs_dir/storage $egs_dir/storage
+  fi
+  sid/nnet3/xvector/get_egs.sh --cmd "$train_cmd" \
+    --nj 8 \
+    --stage 0 \
+    --frames-per-iter 1000000000 \
+    --frames-per-iter-diagnostic 100000 \
+    --min-frames-per-chunk 50 \
+    --max-frames-per-chunk 100 \
```
Comment (amith-ananthram) [inline egs/voxceleb/v1/local/nnet3/xvector/run_meld_xvector.sh:68]: as we're allowing shorter utterances we need to lower the upper bound on the max-frames-per-chunk

File: egs/voxceleb/v1/local/nnet3/xvector/run_meld_xvector.sh [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    --num-repeats 50 \
+    "$data" $egs_dir
+fi
+
+dropout_schedule='0,0@0.20,0.1@0.50,0'
+srand=123
+if [ $stage -eq 9 ]; then
+  steps/nnet3/train_raw_dnn.py --stage=$train_stage \
+    --cmd="$train_cmd" \
+    --trainer.input-model "$input_model" \
```
Comment (amith-ananthram) [inline egs/voxceleb/v1/local/nnet3/xvector/run_meld_xvector.sh:79]: this is where we specify the modified vox speaker id reference model

File: egs/voxceleb/v1/local/nnet3/xvector/run_meld_xvector.sh [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    --trainer.optimization.proportional-shrink 10 \
+    --trainer.optimization.momentum=0.5 \
+    --trainer.optimization.num-jobs-initial=3 \
+    --trainer.optimization.num-jobs-final=8 \
```
Comment (amith-ananthram) [inline egs/voxceleb/v1/local/nnet3/xvector/run_meld_xvector.sh]: todo: reduce the # of jobs here to 2 and 3 respectively
